### DISCUSSION TITLE
CMAKE_PROJECT_TOP_LEVEL_INCLUDES

### DISCUSSION THREAD

**Discussion by jpr42** [microsoft/vcpkg#26681] [General]
The default method `vcpkg` tells CMake users to add the `vcpkg.cmake` module is with a toolchain

```
 cmake -B [build directory] -DCMAKE_TOOLCHAIN_FILE=[path to vcpkg]/scripts/buildsystems/vcpkg.cmake 
```

This works. But I believe a new CMake feature suits this code injection better.

```
 cmake -B [build directory] -D CMAKE_PROJECT_TOP_LEVEL_INCLUDES=[path to vcpkg]/scripts/buildsystems/vcpkg.cmake
```

The main downside is that `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` can't be set as an environment variable. And it was introduced in 3.24

Thoughts?

**Comment by jpr42** [comment]
Link to [CMAKE_PROJECT_TOP_LEVEL_INCLUDES](https://cmake.org/cmake/help/latest/variable/CMAKE_PROJECT_TOP_LEVEL_INCLUDES.html#variable:CMAKE_PROJECT_TOP_LEVEL_INCLUDES) variable.

**Reply by Neumann-A** [reply]
You can probably use that but be wary of `try_compile` configure checks.

**Comment by dg0yt** [comment]
Before CMake 3.24, there is already `CMAKE_PROJECT_INCLUDE`.

IIUC `vcpkg.cmake` is not only code injection, but also actual toolchain setup. I wonder if it could be beneficial to separate these concerns?

**Reply by craigscott-crascit** [reply]
I'm the author of the `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` and dependency provider features in CMake. Both were implemented with specific consideration for how vcpkg works and in consultation with vcpkg leadership. It is my hope that vcpkg will eventually provide a file that users can add to `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` instead of having to use `CMAKE_TOOLCHAIN_FILE` to point at a script vcpkg provides.

I specifically put the `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` injection point after the toolchain file has been read, but before any languages are enabled. This means details about the host and target platform will be available, but toolchain flags (compiler, linker, etc.) have not yet been set. Package managers like vcpkg therefore would have the opportunity to influence those toolchain flags if needed. Users could specify their own toolchain file in the standard CMake way, unlike now where they either have to point `CMAKE_TOOLCHAIN_FILE` at the vcpkg toolchain file and set `VCPKG_CHAINLOAD_TOOLCHAIN_FILE`, or they have to modify their own toolchain file to include the one from vcpkg. If vcpkg provided a file that could be included via `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` instead, users wouldn't have to jump through those hoops.

~~`CMAKE_PROJECT_INCLUDE`~~ `CMAKE_PROJECT_INCLUDE_BEFORE` reads files much earlier. The toolchain file hasn't been read at that point, so none of the target platform details would be defined yet. `CMAKE_PROJECT_INCLUDE` reads files much later. The languages have already been enabled at that point, so compiler flags would have already been determined. Files listed in `CMAKE_PROJECT_INCLUDE_BEFORE` or `CMAKE_PROJECT_INCLUDE` are also read by _every_ `project()` command, not just the first one.

See https://cmake.org/cmake/help/latest/command/project.html#code-injection for details about the sequence of when files are processed by the `project()` command.

**Reply by dg0yt** [reply]
> `CMAKE_PROJECT_INCLUDE` reads files much earlier. The toolchain file hasn't been read at that point, so none of the target platform details would defined yet. 

Are you sure about this particular variable? The documentation states that it is the *last* step of every `project` call, after enabling languages etc., and I use it like that in vcpkg portfiles.
Or do you perhaps refer to `CMAKE_PROJECT_INCLUDE_BEFORE`?

**Reply by dg0yt** [reply]
(And there is also `CMAKE_PROJECT_<PROJECT-NAME>_INCLUDE` which is specific to a particular project and exists even before CMake 3.15.)

**Reply by craigscott-crascit** [reply]
Oh, wait yes, I've updated my comment with the right details. Thanks for catching that.

**Reply by craigscott-crascit** [reply]
So that it isn't lost in the noise, the key point I wanted to make is that `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` is the only variable that inserts the logic at the right place: after the toolchain file has been read, but before languages are enabled.

**Reply by dg0yt** [reply]
Are you aware of particular aspects of `vcpkg.cmake` which do depend on languages being (not) enabled?
I do acknowledge that `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` is close to where `vcpkg.cmake` is injected now, but I wonder if this is really the best injection point.

**Reply by Neumann-A** [reply]
The problem with that variable is that `try_compile` knows nothing about it. So unless that changes vcpkg cannot use it.

**Reply by craigscott-crascit** [reply]
Can you elaborate on why?

**Reply by Neumann-A** [reply]
`try_compile` could run `find_package`. `vcpkg.cmake` changes how `find_package` works. So unless there is some undocumented magic going on with dependencies providers and function overrides `CMAKE_PROJECT_TOP_LEVEL_INCLUDES` is logically not equivalent with setting `CMAKE_TOOLCHAIN_FILE` which gets forwarded to `try_compile`. (The rest of the forwarding is done by CMP137 being NEW)

**Reply by craigscott-crascit** [reply]
Interesting use case. I think you're only talking about whole-project mode of `try_compile()`, yes? I can't see how you could be calling `find_package()` as part of `try_compile()` otherwise. If you have an example of a project that does call `find_package()` within such a scenario, it might be helpful to note that here to provide a concrete motivating example.

I haven't seen this scenario come up before, but it does raise some questions about new features for `try_compile()` coming in CMake 3.25 that may need further discussion. I'll follow up on that specific aspect in CMake's issue tracker (see https://gitlab.kitware.com/cmake/cmake/-/issues/24151).

**Reply by Neumann-A** [reply]
> whole-project mode of try_compile(), yes? 

Yes 

>  it might be helpful to note

I cannot remember which port it was. I can just remember that it was something simple like `find_package(ZLIB|ZSTD|Bzip2|JPEG)` (some compression stuff; some build-in CMake module) probably to test if the compression lib was build with a certain compression feature enabled or to detected an optional dependency. I only stumbled over it since the feature set after configure between release/debug build was different since one of the `try_compiles` yielded a different result between those two configs on windows. The reason the debug build failed the `try_compile` was due to a mismatch of runtime libraries (e.g. tried to link a release lib with debug runtime which fails). Added a `vcpkg-cmake-wrapper` to fix CMakes default behavior and moved on since it wasn't a big deal for me and a common issue/fix within vcpkg.

Be aware that CMake itself as `find_package` in a `try_compile` check https://gitlab.kitware.com/cmake/cmake/-/blob/master/Source/Checks/Curses/CMakeLists.txt#L8
https://gitlab.kitware.com/cmake/cmake/-/blob/master/Source/Checks/Curses.cmake#L21

**Reply by Neumann-A** [reply]
After further thinking it could have been one of the qt modules like, e.g. `qtmultimedia`:
https://github.com/qt/qtmultimedia/blob/531ea0bd937653c0462edf07cae28e473d4b9979/config.tests/alsa/CMakeLists.txt#L16-L18
using `find_package` quite extensively.

**Reply by dg0yt** [reply]
@Neumann-A Is it a problem related also to canonical CMake usage (find modules, config), or strongly connected to vcpkg's use of `find_package` wrappers?

**Reply by Neumann-A** [reply]
> Is it a problem related also to canonical CMake usage (find modules, config), or strongly connected to vcpkg's use of find_package wrappers?

Lets just say `try_compile` was kind of stupid in earlier version of CMake without taking anything into account. The remaining issue has just to do with the use of wrappers in vcpkg correcting cmake behavior. I actually don't currently now if `CMAKE_MODULE_PATH` or other var controlling `find_package()` behavior are forwarded or not (but that might also be left up for the project doing the `try_compile`).

**Reply by craigscott-crascit** [reply]
See [this change](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9532) which aims to address the concerns expressed here, allowing a dependency provider to propagate down into whole-project `try_compile()` calls.
